### PR TITLE
Add benchmarks to all the unfold combinators

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 1]
```diff
@@ -131,57 +410,122 @@ zipWithM size start =
...
     ]
 
-o_1_space_serial_combination :: Int -> [Benchmark]
-o_1_space_serial_combination size =
-    [ bgroup "combination"
-        [ benchIO "zipWithM" $ zipWithM size
-        ]
+
+o_1_space_generation_specialized :: Int -> [Benchmark]
+o_1_space_generation_specialized size =
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:455]: We can have a single benchmark group for "generation" like we have in streams.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 2]
```diff
@@ -131,57 +410,122 @@ zipWithM size start =
...
-        , benchIO "dropWhileMFalse" $ dropWhileMFalse size
-        ]
+o_1_space_generation_resource :: Int -> [Benchmark]
+o_1_space_generation_resource size =
+    [ bgroup
+          "generation/resource"
+          [ benchIO "fromStream" $ fromStream size
+          , benchIO "fromStreamK" $ fromStreamK size
+          -- XXX INVESTIGATE, see the definition
+          -- , benchIO "fromStreamD" $ fromStreamD size
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:435]: I guess this is just going to remain quadratic, we can have this benchmark run for a smaller stream size e.g. "fromStreamD (n/100)" or something like that and remove the `XXX` comment.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 3]
```diff
@@ -131,57 +410,122 @@ zipWithM size start =
...
     ]
 
-o_n_space_serial :: Int -> [Benchmark]
-o_n_space_serial size =
-    [ bgroup "outer-product"
-        [ benchIO "toList" $ toList size
-        , benchIO "toListSome" $ toListSome size
-        ]
+o_1_space_combination :: Int -> [Benchmark]
+o_1_space_combination size =
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:494]: We can call this `o_1_space_zip`, zip is always a separate class in all other benchmarks.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 4]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+_fromStreamD :: Monad m => Int -> Int -> m ()
+_fromStreamD size start =
+    drainGeneration UF.fromStreamD (D.replicate size start)
+
+{-# INLINE _nilM #-}
+_nilM :: Monad m => Int -> Int -> m ()
+_nilM _ start = drainGeneration (UF.nilM return) start
+
+-- XXX The performance is not the best but it is expected because of the way it
+-- is implemented.
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:159]: If this is expected, this should not be a `XXX` comment.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 5]
```diff
@@ -131,57 +410,122 @@ zipWithM size start =
...
-        , benchIO "breakAfterSome" $ breakAfterSome size
-        ]
+o_1_space_transformation :: Int -> [Benchmark]
+o_1_space_transformation size =
+    [ bgroup
+          "transformation"
+          [ benchIO "map" $ map size
+          , benchIO "mapM" $ mapM size
+          , benchIO "mapMWithInput" $ mapMWithInput size
+          , benchIO "takeWhileM" $ takeWhileM size
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:478]: We can make a `o_1_space_filtering` group for all filtering ops, like we have in serial stream benchmarks.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 6]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+    drainTransformationDefault size (UF.mapM (return . (+) 1)) start
+
+{-# INLINE mapMWithInput #-}
+mapMWithInput :: Monad m => Int -> Int -> m ()
+mapMWithInput size start =
+    drainTransformationDefault
+        size
+        (UF.mapMWithInput (\a b -> return $ a + b))
+        start
+
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Unfold.hs:284]: Add a section header here to demarcate the "Filtering" section.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 7]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+    => Int
+    -> (Unfold m Int Int -> Unfold m Int Int -> Unfold m e f)
+    -> e
+    -> m ()
+drainProductDefault size = drainProduct src1 src2
+
+    where
+
+    src1 = UF.take size UF.enumerateFromIntegral
+    src2 = UF.take size (UF.supplySecond UF.enumerateFromStepIntegral 2)
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs:77]: Oh yeah, don't know why I did this. But the benchmarks don't seem to care anyway.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 8]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+    drainTransformationDefault size (flip UF.supply start) undefined
+
+
+{-# INLINE supplyFirst #-}
+supplyFirst :: Monad m => Int -> Int -> m ()
+supplyFirst size start =
+    drainTransformation
+        (UF.take size UF.enumerateFromStepIntegral)
+        (flip UF.supplyFirst start)
+        1
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs:104]: I'm benchmarking `supplyFirst`. Need the seed to be `(a, b)`. I could do an `lmap` I guess.

[the remark this answers is not in the document]
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs:100]: We can leave it like such for now.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 9]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+takeWhile size start =
+    drainTransformationDefault
+        size
+        (UF.takeWhile (\b -> b <= size + start))
+        start
+
+{-# INLINE filter #-}
+filter :: Monad m => Int -> Int -> m ()
+filter size start =
+    drainTransformationDefault size (UF.filter (\_ -> True)) start
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs]: We can leave it like this for now. This is shorter than `Prelude.const`

File: benchmark/Streamly/Benchmark/Data/Unfold.hs [hunk 10]
```diff
@@ -11,118 +11,397 @@ module Main (main) where
...
+concatMapM size start = drainUnfold (UF.concatMapM unfoldInGen unfoldOut) start
+
+    where
+
+    sizeOut = 100
+    sizeIn = size `div` sizeOut
+
+    unfoldInGen i =
+        return
+            $ UF.lmap (\() -> undefined)
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs:395]: We can do this, but probably in another set of changes. After this and #722 is merged

**Comment by adithyaov** [comment]
This does not the changes of `iterateM` yet. Once this is approved, the batch PR will have the required changes.

File: benchmark/Streamly/Benchmark/Data/Unfold.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply benchmark/Streamly/Benchmark/Data/Unfold.hs:395]: Can we raise an issue so that we do not lose this?

**Review by harendra-kumar — APPROVED** [review]

**Comment by adithyaov** [comment]
See #735

**Pull request closed without merging**